presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsAnnotationProcessor.java

```diff
@@ -0,0 +1,243 @@
+    }
+
+    private static double processSingleArgumentStatistic(
+            double outputRowCount,
+            double nullFraction,
+            CallExpression callExpression,
+            List<Double> sourceStats,
+            int sourceStatsArgumentIndex,
+            StatsPropagationBehavior operation)
+    {
+        // sourceStatsArgumentIndex is index of the argument on which
+        // ScalarPropagateSourceStats annotation was applied.
+        double statValue = NaN;
+        if (operation.isMultiArgumentStat()) {
+            for (int i = 0; i < sourceStats.size(); i++) {
+                if (i == 0 && operation.isSourceStatsDependentStats() && isFinite(sourceStats.get(i))) {
+                    statValue = sourceStats.get(i);
+                }
+                else {
+                    switch (operation) {
+                        case MAX_TYPE_WIDTH_VARCHAR:
+                            statValue = returnNaNIfTypeWidthUnknown(getTypeWidthVarchar(callExpression.getArguments().get(i).getType()));
+                            break;
+                        case USE_MIN_ARGUMENT:
+                            statValue = min(statValue, sourceStats.get(i));
+                            break;
+                        case USE_MAX_ARGUMENT:
+                            statValue = max(statValue, sourceStats.get(i));
+                            break;
+                        case SUM_ARGUMENTS:
+                            statValue = statValue + sourceStats.get(i);
+                            break;
+                        case SUM_ARGUMENTS_UPPER_BOUNDED_TO_ROW_COUNT:
+                            statValue = min(statValue + sourceStats.get(i), outputRowCount);
+                            break;
+                    }
+                }
+            }
+        }
+        else {
+            switch (operation) {
+                case USE_SOURCE_STATS:
+                    statValue = sourceStats.get(sourceStatsArgumentIndex);
+                    break;
+                case ROW_COUNT:
+                    statValue = outputRowCount;
+                    break;
+                case NON_NULL_ROW_COUNT:
+                    statValue = outputRowCount * (1 - firstFiniteValue(nullFraction, 0.0));
+                    break;
+                case USE_TYPE_WIDTH_VARCHAR:
+                    statValue = returnNaNIfTypeWidthUnknown(getTypeWidthVarchar(callExpression.getArguments().get(sourceStatsArgumentIndex).getType()));
+                    break;
+            }
+        }
+        return statValue;
+    }
+
+    private static int getTypeWidthVarchar(Type argumentType)
+    {
+        if (argumentType instanceof VarcharType) {
+            if (!((VarcharType) argumentType).isUnbounded()) {
+                return ((VarcharType) argumentType).getLengthSafe();
+            }
+        }
+        return -VarcharType.MAX_LENGTH;
+    }
+
+    private static double returnNaNIfTypeWidthUnknown(int typeWidthValue)
```

Comment:
I'd be tempted to just use a negative value here. Is it realy necessary to use NaN?

Reply:
Yes, NaN indicates unknown.